subsys/power/device_pm.c

```diff
@@ -0,0 +1,226 @@
+		dev->pm->status = DEVICE_PM_REQ_STAT_DONE;
+	}
+out:
+	/* Notify the driver through alert */
+	if (dev->pm->alert.handler) {
+		k_alert_send(&dev->pm->alert);
+	}
+	k_sem_give(&dev->pm->lock);
+}
+
+/*
+ * Set the device PM state to target PM state based on
+ * the current state. If the PM request is of Async type
+ * then a work item will be submitted to PM worker queue
+ * else it will call device_set_power_state() to set the
+ * device PM state to target state.
+ */
+static int device_pm_set_state(struct device *dev,
+			       u32_t target_state, u32_t pm_flags)
+{
+	int ret;
+	u32_t current_state;
+
+	/* Get current device power state */
+	ret = device_get_power_state(dev, &current_state);
+	if (ret) {
+		LOG_ERR("Get power state error(%d)\n", ret);
+		goto out;
+	}
+
+	/* Return if current state is same as target state */
+	if (current_state == target_state) {
+		goto out;
+	}
+
+	/* Cancel already queued work */
+	if (dev->pm->status == DEVICE_PM_REQ_STAT_INIT) {
+		k_delayed_work_cancel(&dev->pm->dwork);
+		dev->pm->status = DEVICE_PM_REQ_STAT_CANCEL;
+		/* Notify the driver through alert */
+		if (dev->pm->alert.handler) {
+			k_alert_send(&dev->pm->alert);
+		}
+	}
+
+	/* Schedule the worker incase of async request */
+	if (pm_flags & DEVICE_PM_ASYNC) {
+		dev->pm->request = target_state;
+		dev->pm->status = DEVICE_PM_REQ_STAT_INIT;
+		k_delayed_work_submit_to_queue(&devpm_workq,
+					&dev->pm->dwork, dev->pm->timeout);
+	} else {
+		ret = device_set_power_state(dev, target_state);
+		if (ret) {
+			LOG_ERR("Set power state(%d) error(%d)\n",
+							target_state, ret);
+		}
+	}
+
+out:
+	return ret;
+}
+
+static int device_pm_handle_get(struct device *dev, u32_t pm_flags)
+{
+	int ret;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (dev->pm->usage >= UINT_MAX) {
+		LOG_ERR("usage count overflow error!!\n");
+		ret = -ERANGE;
+		goto get_out;
+	}
+	dev->pm->usage++;
+
+	if (!dev->pm->enable) {
+		ret = 0;
+		goto get_out;
+	}
+
+	ret = device_pm_set_state(dev,
+			DEVICE_PM_ACTIVE_STATE, pm_flags);
+
+get_out:
+	k_sem_give(&dev->pm->lock);
+	return ret;
+}
+
+static int device_pm_handle_put(struct device *dev, u32_t pm_flags)
+{
+	int ret;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (dev->pm->usage <= 0) {
+		LOG_ERR("usage count underflow error!!\n");
+		ret = -ERANGE;
+		goto put_out;
+	}
+	dev->pm->usage--;
+
+	if (!dev->pm->enable) {
+		ret = 0;
+		goto put_out;
+	}
+
+	ret = device_pm_set_state(dev,
+			DEVICE_PM_SUSPEND_STATE, pm_flags);
+
+put_out:
+	k_sem_give(&dev->pm->lock);
+	return ret;
+}
+
+int device_pm_get(struct device *dev)
+{
+	return device_pm_handle_get(dev, DEVICE_PM_ASYNC);
+}
+
+int device_pm_get_sync(struct device *dev)
+{
+	return device_pm_handle_get(dev, 0);
+}
+
+int device_pm_put(struct device *dev)
+{
+	return device_pm_handle_put(dev, DEVICE_PM_ASYNC);
+}
+
+int device_pm_put_sync(struct device *dev)
+{
+	return device_pm_handle_put(dev, 0);
+}
+
+void device_pm_enable(struct device *dev)
```

Comment:
IMHO there are 2 use-cases missed:
- Device with PM disabled, not used: After `device_pm_enable()` should be suspended.
- Device with PM enabled, not used: After `device_pm_disable()` should be woken up.

Reply:
Ok got it. Thanks.